Fix null key removal in nested subcharts

jamoflaw opened the pull request:
closes #9136
closes #9027

This PR contains two commits, the first replicates the behaviour described in #9136 and the second resolves it.

The original unit-tests did have a case for the subchart values and it was passing, so I spent some time trying to work out why there is a difference between the behaviour of the unit-tests and real world usage. It seems that the bug is related to the way the chart object is mutated by chartutil.ProcessDependencies. During that call the chart objects values are merged with its subcharts leading to a situation where all values are merged together into the parent chart's values.

To replicate this behaviour in the unit tests I just added the following to the parent charts definition:

```
			"pequod":   map[string]interface{}{
				"ahab":     map[string]interface{}{
					"scope": "ahab",
					"nested": map[string]interface{}{
						"foo": true,
						"bar": true,
					},
				},
			},
```

Since `bar` is now defined in that parent chart object, when the call to `coalesceValues(printf, ch, dest, prefix)` was made, the dest object was mutated and null keys were removed from both the parent chart _and_ subchart values. The subsequent call to `coalesceDeps(printf, ch, dest, prefix)` then has no more null keys anymore so the subcharts no longer have any keys removed, which causes the subcharts values to be merged _back_ into the parent chart (undo-ing the null removal from the previous call to coalesceValues).

The changes made effectively delay the removal of the null keys until after both the `coalesceValues` and `coalesceDeps` calls are completed.

- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

sabre1041 commented:
@jamoflaw I attempted the following test:

subchart with values

```yaml
resources:
  limits:
    cpu: 100m
    memory: 128Mi
  requests:
    cpu: 100m
    memory: 128Mi
```

chart that uses subchart as a dependency:

```yaml
subchart:
  resources:
    requests:
      cpu: 1000m
      memory: 2Gi
    limits:
      cpu: null
      memory: 2Gi
```

Resulted in the following to be rendered:

```yaml
resources:
  limits:
    cpu: 100m
    memory: 2Gi
  requests:
    cpu: 1000m
    memory: 2Gi
```

Unfortunately, the original value remains

jamoflaw commented:
@sabre1041 do you happen to have the chart you tested it on available? I tested this PR on one of my internal ones and it was able to remove mapping values which were set in the subchart from a values file. (ie `-f`) Were you overriding them in the parent chart? I wonder if there are more issues than I originally found. 

So my testing was effectively the equivalent of sub chart with values:

```
resources:
  limits:
    cpu: 100m
    memory: 128Mi
  requests:
    cpu: 100m
    memory: 128Mi 
```

Then in the parent chart - nothing above is set

Then I am passing in (with `-f`) an overrides file with:

```
subchart:
  resources:
    limits:
      cpu: null
      memory: 128Mi
    requests:
      cpu: 100m
      memory: 128Mi 
```

Which results in:

```
subchart:
  resources:
    limits:
      memory: 128Mi
    requests:
      cpu: 100m
      memory: 128Mi 
````

sabre1041 commented:
@jamoflaw this is unfortunately a large issue where there are multiple scenarios that need to be accounted for. 

My test did not explicitly reference an external values file but were referenced in the parent chart  that used the chart as a dependency

jamoflaw commented:
Agreed! I've written another unit test which covers that scenario and I'll have a look at seeing whats going on

jamoflaw commented:
OK yeah I've replicated it - the same case works in the current unit tests though, which isn't ideal!! I'll update the PR with a fix and write a unit test which triggers it. Thanks @sabre1041

jamoflaw commented:
So this update fixes it for parent chart overrides too. The issue seems to be that CoalesceValues isn't idempotent when it comes to null removal behaviour and it is called a LOT throughout the rendering pipeline basically removing nulls on the first call then remerging values back in from subcharts on subsequent calls.

What I did here was split the CoalesceValues into two (keeping the old names pointing back at the new variant which shares its old behaviour) so it is clear what each variant does, and then delay the removal of null keys _even further_ than previous (basically at the end of the rendering pipeline)

The two variants are *KeepNil and *TrimNil which as their names suggest either keep nils around or trim them. Where possible the Trim variant is used to make sure behaviour isn't needlessly changed, but during initial dependency gathering its changed to the KeepNil variant to avoid early stripping of null keys.

jamoflaw committed:
Stop CoalesceValues from trimming nil values from
the values. Creating new public function TrimNilValues
which does the final nil value removal. This is called in
ToRenderValues before final schema validation.

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw commented:
@sabre1041 as an alternative to this PR I have also prepared https://github.com/helm/helm/compare/main...jamoflaw:helm:fix-null-removal-in-nested-subcharts-alt. Both now handle my original case (where overrides in files provided with `-f` couldn't override nested null keys) and the case you mentioned where a parent chart defines a null key value to remove a subchart key.

Personally I feel the alternative branch above is a simpler and more effective solution _but_ unlike this PR it means that the current behaviour of CoalesceValues changes _slightly_. The current behaviour is inconsistently buggy around null removal whereas in my alternative branch above it is always consistently now _not_ removing null keys. Instead this is delegated to a new public method `TrimNilValues` which has the sole purpose of achieving the removal of null keys. This null removal is then done late in the pipeline in `ToRenderValues`.

I have purposely not yet raised this as a new PR as wanted to get some feedback on if this approach is viable given it alters the CoalesceValues behaviour slightly.

sabre1041 commented:
@jamoflaw can confirm that the PR above does work as expected with the use case that I brought up. I'll look more into the implementation and any potential implication of the change in functionality within the `CoalesceValues` function

Vermyndax commented:
Any hope of this one getting pushed through for the 11/09 release? Any way I can upvote it?

sabre1041 commented:
@Vermyndax I can take a look again later on this week

jamoflaw commented:
For my 2c on the issue:

I prefer this implementation https://github.com/helm/helm/compare/main...jamoflaw:helm:fix-null-removal-in-nested-subcharts-alt as its significantly cleaner IMHO, but replaces a buggy behaviour - where null keys are removed inconsistently, with a consistent behaviour where null keys are instead retained _and_ returned by that function. Internally then keys are removed with a second dedicated function.

I'm happy to rewrite as required to get this one working, maybe taking the alternative approach PR and making it backward compatible, splitting new behaviour into new functions? This does hit us quite a bit as we use nested helm charts a _lot_, but would like some guidance on the approach just so I don't rewrite it in the wrong direction.

Vermyndax commented:
...longingly came by to check on all of you... hope you're well... :) Is there anything I can do to help push this along?

joelio commented:
Hey, is there anything to help expedite this, it'll be super useful for our work and the sooner the better for us!

jamoflaw commented:
> For my 2c on the issue:
> 
> I prefer this implementation [main...jamoflaw:helm:fix-null-removal-in-nested-subcharts-alt](https://github.com/helm/helm/compare/main...jamoflaw:helm:fix-null-removal-in-nested-subcharts-alt) as its significantly cleaner IMHO, but replaces a buggy behaviour - where null keys are removed inconsistently, with a consistent behaviour where null keys are instead retained _and_ returned by that function. Internally then keys are removed with a second dedicated function.
> 
> I'm happy to rewrite as required to get this one working, maybe taking the alternative approach PR and making it backward compatible, splitting new behaviour into new functions? This does hit us quite a bit as we use nested helm charts a _lot_, but would like some guidance on the approach just so I don't rewrite it in the wrong direction.

Just a follow up on this comment - I will have time to look at it this weekend - does it seem like something which would be worth following up on as I wouldn't want this issue to become stale

sabre1041 commented:
@jamoflaw I'd like to move this forward. please take a look at it at your convenience and we can push forward as it is an issue that does need to be addressed

kladiv commented:
+1 i follow this PR, useful in scenario when we want to remove (default) _resources.limits_ of `kube-prometheus-stack` helm chart (made of subcharts)

tonyqui commented:
@hickeyma, any clue why it didn't fit milestone 3.10.3?
Do you see still anything missing?

hickeyma commented:
@tonyqui It is because v3.10.3 is a security release.

brsolomon-deloitte commented:
> Just a follow up on this comment - I will have time to look at it this weekend

@jamoflaw we appreciate you getting this one to the finish line.

Commit:
Merge remote-tracking branch 'jamoflaw/fix-null-removal-in-nested-subcharts-alt' into main

jamoflaw committed:
Ensure MergeValues is called where CoalesceValues was previously. The
publicly exposed function CoalesceValues is now backward compatible.

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw commented:
I've just force pushed a new set of commits which are functionally equivalent to the original PR, but cleaner and maintain existing behaviour on the exposed public functions:

Creates a new function MergeValues which maintains nil values
Creates a new function TrimNilValues which removes nil values from maps
Maintain the backward compatible behaviour of the public function CoalesceValues by combining MergeValues followed by TrimValues

I have included some new tests for MergeValues (which pretty much replicate the original CoalesceValues except expecting the nils to be maintained)

Can I ask for comment on the `MergeValues`/`CoalesceValues` split? Is maintaining the potentially buggy behaviour worth the split or should CoalesceValues be allowed to return nils?

jamoflaw commented:
Testing on the problematic charts on my side seems good - its correctly removing the cpu limits, when requested to, on the Elastic chart https://github.com/elastic/helm-charts/blob/main/elasticsearch/values.yaml#L95 which is where I notice it most often.

Let me know if there are any issues - fingers crossed this one can be put to bed!

jamoflaw commented:
FYI @sabre1041 incase it was missed

jamoflaw commented:
Just a quick bump incase anyone is able to look at this!

@sabre1041 @joejulian

sabre1041 reviewed (approved):
LGTM

sabre1041 commented:
@jamoflaw code looks good

could you produce a new workflow run so that we can get green across the board as the previous workflow run?

seanmorton commented:
> @jamoflaw code looks good
> 
> could you produce a new workflow run so that we can get green across the board as the previous workflow run?

Is this referring to the checks? Each seem to be green - anything I can do to help this along?

jamoflaw commented:
> > @jamoflaw code looks good
> > could you produce a new workflow run so that we can get green across the board as the previous workflow run?
> 
> Is this referring to the checks? Each seem to be green - anything I can do to help this along?

Yeah this was just referencing the checks when I originally posted which were not complete. They all look good now! Thanks whoever sorted that part!

kolorful commented:
Is there anything else needed for this MR to be merged? https://github.com/helm/helm/blob/main/CONTRIBUTING.md does seem to indicate PRs need two maintainers' approval for merging?

jamoflaw committed:
Merge branch 'main' into fix-null-removal-in-nested-subcharts

Signed-off-by: jamoflaw <jamoflaw@gmail.com>

jamoflaw commented:
> @jamoflaw hate to do this, but can you resolve the conflicting files.
> 
> Working in the background to get the necessary eyes on this PR so that it can be integrated. Appreciate everyone's patience

Thanks @sabre1041 - conflicts should be fixed now

jamoflaw committed:
Fix whitespace issues reported by gofmt

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw commented:
OK I'm hoping that will fix the formatting issues introduced by using the github UI! _sigh_

sofam commented:
Any progress on getting this merged in? I just hit this issue with multiple of our Helm charts.

jamoflaw commented:
Is there anything I need to do at this stage? Do we need additional approvers given the size/L tag @sabre1041?

gjenkins8 reviewed (commented):
Non-maintainer (community member), but had a few minorish comments.

Otherwise, agree, we document null-key removal [here](https://helm.sh/docs/chart_template_guide/values_files/#deleting-a-default-key), and this should apply correctly to subcharts

jamoflaw commented on pkg/chartutil/values.go:
```diff
@@ -151,11 +151,14 @@ func ToRenderValues(chrt *chart.Chart, chrtVals map[string]interface{}, options
 	}
 
+	// Remove nil values from the coalesced map
+	vals = TrimNilValues(vals)
```
Yup its basically done last as the previous issue was due to CoalesceValues basically being non-idempotent but being called as if it was. This delays that call to the last moment before its needed

gjenkins8 replied:
nice, thanks for clarifying.

jamoflaw commented on pkg/chartutil/coalesce.go:
```diff
@@ -39,15 +39,46 @@ func concatPrefix(a, b string) string {
 //     the values destined for its dependencies.
+//
+//     Note that CoalesceValues automatically removes nil values
 func CoalesceValues(chrt *chart.Chart, vals map[string]interface{}) (Values, error) {
```
I think it could actually - currently `TrimNilValues` doesn't return the err on line 96 which I think could be improved as well, if that was done the sigatures should be compatible.

Ill give this a go tomorrow and test it out

jamoflaw replied:
So it works out as:

```
func CoalesceValues(chrt *chart.Chart, vals map[string]interface{}) (Values, error) {
	valsCopy, err := MergeValues(chrt, vals)
	if err != nil {
		return vals, err
	}
	return TrimNilValues(valsCopy)
}
```

to avoid passing the error down into `TrimNilValues` - I'll run against my larger charts which show the original error tomorrow morning

mattfarina commented on cmd/helm/status.go:
```diff
@@ -191,7 +191,7 @@ func (s statusPrinter) WriteTable(out io.Writer) error {
 		fmt.Fprintln(out)
 
-		cfg, err := chartutil.CoalesceValues(s.release.Chart, s.release.Config)
+		cfg, err := chartutil.MergeValues(s.release.Chart, s.release.Config)
```
This section of code is used by `helm get all`, `helm --debug install`, and possibly other places. Before this change values that were `null` were not displayed in the output. After this change they are displayed with a value of `null` in the computed values. Is this intentional?

rouke-broersma replied:
Imo that is what the output should explicitly show, because that is the whole purpose of this pr. Nill values should be propagated. So any result output should list that this is what is happening. Explicitly.

jamoflaw replied:
This is what I was mentioning here https://github.com/helm/helm/pull/11440#issuecomment-1280070498

Basically the original behaviour of the public function `CoalesceValues` is maintained in this PR and a new public function `MergeValues` is also created which does not trim nil values from the resulting output. This is to allow those nils to persist until the last call to TrimNilValues where they will be removed.

In short:
* MergeValues is effectively a "fixed" CoalesceValues - but it leaves nulls in place and returns them
* CoalesceValues maintains its previous _potentially_ buggy behaviour but _doesn't_ return null ever - I say potentially as its only when its called in succession as helm does internally - externally this behaviour may not have been seen

I have made the assumtion in the original comment that CoalesceValues should never return null - this may not be the case and in which case we could simplify this PR and the potential annoyance of wording between Merge and Coalesce - but it is potentially changing external behaviour which this current PR fixes albeit with the aforementioned naming challenges.

Before I make additional changes its probably worth clearing up if my assumption is valid or if we can make CoalesceValues return nulls

mattfarina reviewed (commented):
Thanks for the PR and for putting work into this, @jamoflaw 

I did a brief run though and have some questions (and one comment). A big thing is to look at all the consequences of the change. Including how output changes.

I'll circle back with more feedback asap but wanted to get these questions added before I walk away from the PR for the day.

jamoflaw commented on pkg/chartutil/coalesce.go:
```diff
@@ -39,15 +39,46 @@ func concatPrefix(a, b string) string {
+//     the values destined for its dependencies.
+//     Note that MergeValues does NOT automatically remove nil values, it is up
+//     to the caller to ensure that TrimNilValues is called at the appropriate time
+func MergeValues(chrt *chart.Chart, vals map[string]interface{}) (Values, error) {
```
Its the opposite so MergeValues is _called_ internally but leaves null values as is - CoalesceValues maintains exactly the same external behaviour - in that nulls are removed albeit maintaining that incorrect behaviour if its called multiple times such as how Helm internally does it

jamoflaw commented on pkg/chartutil/coalesce.go:
```diff
@@ -56,6 +87,28 @@ func CoalesceValues(chrt *chart.Chart, vals map[string]interface{}) (Values, err
 	return coalesce(log.Printf, chrt, valsCopy, "")
 }
 
+// TrimNilValues removes nil/null values from maps.
+//
+//	To be used after CoalesceValues to remove keys as needed
```
Ah this is true it should mention MergeValues here instead

alice-sawatzky commented:
any updates? @jamoflaw if you're burned out on the back and forth i'd be happy to pick this up

prune998 commented:
@alice-sawatzky you should take over as this is 3 years old now and we still can't "remove" value in a subchart... :/

jamoflaw commented:
> @alice-sawatzky you should take over as this is 3 years old now and we still can't "remove" value in a subchart... :/

I'm happy to move this forward but was waiting on feedback from the questions I've asked a couple times

https://github.com/helm/helm/pull/11440#issuecomment-1280070498
https://github.com/helm/helm/pull/11440#discussion_r1180785475

Specifically on direction - do we _need_ to split these functions or can this PR be simplified as we could return null from the original function?

jamoflaw commented:
> > @alice-sawatzky you should take over as this is 3 years old now and we still can't "remove" value in a subchart... :/
> 
> I'm happy to move this forward but was waiting on feedback from the questions I've asked a couple times
> 
> [#11440 (comment)](https://github.com/helm/helm/pull/11440#issuecomment-1280070498) [#11440 (comment)](https://github.com/helm/helm/pull/11440#discussion_r1180785475)
> 
> Specifically on direction - do we _need_ to split these functions or can this PR be simplified as we could return null from the original function?

basically I don't want to unnecessarily make this PR more complex than it needs to be - specifically changing the language used for the function names - its an annoying change to have to make - and I'm seeking advice on if that is wanted

mattfarina commented:
I have a few things to add here...

First, we can't change the behavior of a public function. If it removed null values before and that behavior is removed from it, that's a problem because it breaks SemVer. People use the Helm SDK and we need to avoid breaking changes until Helm v4 comes out. It appears that means the "simpler change" won't work.

I'm skipping the second question because I was looking at #9027 and found the difference is in where `ProcessDependencies` executes. When it runs it alters the parent charts `Values` property to import the child chart values. I'm looking through that part of the process for a possible bug.

jamoflaw commented:
> First, we can't change the behavior of a public function. If it removed null values before and that behavior is removed from it, that's a problem because it breaks SemVer. People use the Helm SDK and we need to avoid breaking changes until Helm v4 comes out. It appears that means the "simpler change" won't work.

Yeah that does make sense

Ref `ProcessDependencies` I tried:

```
	newvals, _ := copystructure.Copy(vals)
	if err := chartutil.ProcessDependencies(chrt, newvals.(map[string]interface{})); err != nil {
		return nil, err
	}
```

to ensure it works with a copy and it didn't seem to make any difference.

I did notice though that the chart.Values value gets worked on a lot in those calls via a ptr and that passes through the same CoalesceValues functions multiple times. It seems that some of those calls are intended to mutate that object though?

mattfarina commented:
@jamoflaw I'm fairly deep into this one now. I might have a fix. More testing required. I found two problems...

1. When processing dependencies there is a place where the values are being manipulated. A copy of them fixes that issue.
2. Coalescing the values removes null values. During the dependency phase this happens. Later, they are coalesced again but without the null values. A merge that does not remove nulls may handle this situation. Creating merge here might be a little more complicated than trimming the nulls. There are times, in the current code and tests, that they are left and the "delete" step is skipped.

I'm testing some ideas out right now to better deal with the 2 issues. Also need to make the changes in a backward compatible manner. I'll have more to share soon.

mattfarina commented:
@jamoflaw Can you take a look at https://github.com/helm/helm/pull/12162? This is my attempt to fix this and some other related problems that needed to be fixed to make this all work.

jamoflaw commented:
Thanks @mattfarina - should this PR be closed up with a link to the new one in that case?

joejulian commented:
Based on my tests, the issues attached have both been fixed by https://github.com/helm/helm/pull/12162. Closing.

mattfarina commented:
@shazada a fix for this problem is scheduled for release on September 13th.
